e2xgrader/apps/api.py

```diff
@@ -140,3 +144,187 @@ def get_task_submissions(self, assignment_id, notebook_id, task_id):
             submission['index'] = idx
 
         return submissions
+
+class E2X_Gradebook(Gradebook):
+
+    def update_grade(self, grade_cell: str, notebook: str, assignment: str, student: str, **kwargs: dict) -> None:
+        """Updates grade for a submissions for a particular cell in a particular notebook in an assignment.
+
+        Arguments
+        ---------
+        grade_cell: string
+            The name of the grade cell
+        notebook_id: string
+            The name of the notebook
+        assignment_id: string
+            The name of the assignment
+        student: string
+            The id of the student
+        Returns
+        -------
+        None
+        """
+        try:
+            grade = self.find_grade(grade_cell, notebook, assignment, student)
+        except MissingEntry:
+            self.logger.warning('No grade found!')
+        else:
+            for attr in kwargs:
+                setattr(grade, attr, kwargs[attr])
+            try:
+                self.db.commit()
+            except (IntegrityError, FlushError) as e:
+                self.logger.warning('Commit to database failed!')
+                self.db.rollback()
+                raise InvalidEntry(*e.args)
+
+        return None
+
+    def list_updated_cells(self, notebook: str, assignment: str) -> dict:
+        """Lists the updated autograde cell id and content from the source directory.
+
+        Arguments
+        ---------
+        notebook: string
+            The name of the notebook
+        assignment: string
+            The name of the assignment
+        Returns
+        -------
+        updated_cells: list
+            Dictionary where the keys are the ids of the cell and the values are the content.
+        """
+        nb = nbformat.read("source/" + assignment + "/" + notebook + ".ipynb", as_version = 4)
```

Comment:
Use the coursedir instance here to find the correct path (see for example https://github.com/jupyter/nbgrader/blob/c04f9c497eed5e481b57e557ef61989e073c5415/nbgrader/coursedir.py#L122). Join path with os.path.join, not with string concatenation

Reply:
Changed to the following code.
os.path.join(CourseDirectory().source_directory, assignment, notebook + '.ipynb')
The only way to add the extension '.ipynb' was to use concatenation.
Could you check if there is another way?